src/goto-analyzer/show_callsites.cpp

```diff
@@ -0,0 +1,101 @@
+/*******************************************************************\
+
+Module: Write a function call graph to a CSV file
+
+Author: Norbert Manthey, nmanthey@amazon.de
+
+\*******************************************************************/
+
+#include "show_callsites.h"
+
+#include <goto-programs/compute_called_functions.h>
+#include <goto-programs/goto_model.h>
+#include <util/cprover_prefix.h>
+#include <util/json.h>
+#include <util/prefix.h>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+/** show_callsites writes a call graph to CSV
+ *
+ * This function creates a map between functions and consumed functions, and
+ * stores this relationship in a CSV file, to be processed outside of
+ * goto-analyzer lateron.
+ *
+ * Functions, whose address is consumed, are listed in a separate section so
+ * that outside reasoning has visibility as well.
+ **/
+
+void show_callsites(
+  const goto_modelt &goto_model,
+  const std::string &output_filename)
+{
+  // generate all functions which we cannot be sure where they are called
+  // because their address is consumed, and stored into function pointers
+  std::unordered_set<irep_idt> excluded_functions;
+  compute_address_taken_functions(
+    goto_model.goto_functions, excluded_functions);
+
+  std::ofstream output_file;
+
+  if(output_filename != "-")
+  {
+    output_file.open(output_filename);
+    if(!output_file.is_open())
+      throw "failed to open callsite output file";
+  }
+
+  std::ostream *output_stream =
+    output_filename == "-" ? &std::cout : &output_file;
+
+  // handle functions whose address is used
+  json_objectt result;
+
+  json_arrayt by_address;
+  for(const auto &name : excluded_functions)
+  {
+    const std::string str_name = id2string(name);
+    if(
+      has_prefix(str_name, CPROVER_PREFIX) ||
+      (str_name.find('$') != std::string::npos))
+      continue;
+
+    by_address.push_back(json_stringt(str_name));
+  }
+  if(by_address.size())
+    result["__BY_ADDRESS__"] = by_address;
+
+  // go through call graph
+  int functions = 0;
+  forall_goto_functions(f_it, goto_model.goto_functions)
+  {
+    // ignore internal functions
+    const std::string str_name = id2string(f_it->first);
+    if(
+      has_prefix(str_name, CPROVER_PREFIX) ||
+      (str_name.find('$') != std::string::npos))
+      continue;
+
+    json_arrayt function_line;
+    const goto_programt &program = f_it->second.body;
+    forall_goto_program_instructions(i_it, program)
+    {
+      if(!i_it->is_function_call())
+        continue;
+
+      const auto &function = to_code_function_call(i_it->code).function();
+      auto &se = to_symbol_expr(function).get_identifier();
+      if(id2string(se).find('$') != std::string::npos)
+        continue;
+
+      function_line.push_back(json_stringt(id2string(se)));
+    }
+
+    functions++;
+    result[id2string(f_it->first)] = function_line;
```

Comment:
Keys should be camelCase. Hm... I see now... That's a bit atypical for a JSON file. Typically it should be rather structured as follows:
```
[
  {
     "name" : "my_caller",
     "callees" : [ "my_callee1", "my_callee2" ]
  },
  ...
]
```

Reply:
I would like to stick to the "caller" -> callees structure, as that represents the graph that this structure represents. I already moved away from CSV, i want to keep at least "have each adjacency list in a single line, starting with the caller".

Reply:
I agree with @nmanthey. Let's store map as a map.